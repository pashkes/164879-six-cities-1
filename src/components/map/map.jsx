import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import CitiesCoordinates from './../../mock/coordinatesCities';

class Map extends PureComponent {
  constructor(props) {
    super(props);
    this._zoom = 12;
    this._marker = {
      iconUrl: `img/marker.svg`,
      iconSize: [30, 30]
    };
    this._group = null;
  }

  __addMarkers() {
    this._group = this.props.leaflet.layerGroup().addTo(this._map);
    for (let item of this.props.coordinates) {
      this.props.leaflet
        .marker(item, this._marker)
        .addTo(this._group);
    }
  }

  _initMap() {
    const leaflet = this.props.leaflet;
    const city = CitiesCoordinates.get(this.props.currentCity);
    this._map = leaflet.map(`map`, {
      center: city,
      zoom: this._zoom,
      zoomControl: false,
      marker: true
    });
    this._map.setView(city, this._zoom);
    leaflet
      .tileLayer(`https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`, {
        attribution: `&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors &copy; <a href="https://carto.com/attributions">CARTO</a>`
      })
      .addTo(this._map);
    this.__addMarkers();
  }

  componentDidMount() {
    this._initMap();
  }

  componentDidUpdate() {
    this._group.clearLayers();
    this._map.setView(CitiesCoordinates.get(this.props.currentCity), this._zoom);
    for (let item of this.props.coordinates) {
      this.props.leaflet
        .marker(item, this._marker)
        .addTo(this._group);
    }
  }

  render() {
    return (
      <div className="cities__map" id="map"/>
    );
  }
}

Map.propTypes = {
  coordinates: PropTypes.array.isRequired,
  leaflet: PropTypes.object,
  currentCity: PropTypes.string.isRequired
};

const mapStateToProps = (state) => ({
  currentCity: state.city,
  coordinates: state.offers.filter((offer) => {
    return offer.city === state.city;
  }).map((it) => it.coordinates)
});


export default connect(mapStateToProps)(Map);
