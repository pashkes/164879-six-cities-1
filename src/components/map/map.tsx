import * as React from "react";
import * as leaflet from "leaflet";

import {CITIES} from "./../../constants";

const MapConfig = {
  ID: `map`,
  ZOOM: 12,
  MARKER_PATH: `img/marker.svg`,
  ACTIVE_MARKER_PATH: `img/active-marker.svg`,
  MARKER_SIZE: [30, 30],
  TITLE_LAYER: `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`
};

interface Props {
  coordinates: [[number, number]],
  currentCity: string,
  selectedOffer: [number, number],
}

class Map extends React.PureComponent<Props> {
  private map: any;
  private group: any;
  private icon: any;
  private activeIcon: any;

  constructor(props) {
    super(props);

    this.group = null;
    this.icon = leaflet.icon({
      iconUrl: MapConfig.MARKER_PATH,
      iconSize: [30, 30],
    });
    this.activeIcon = leaflet.icon({
      iconUrl: MapConfig.ACTIVE_MARKER_PATH,
      iconSize: [30, 30],
    });
  }

  _addMarkers() {
    const {selectedOffer, coordinates} = this.props;

    this.group = leaflet.layerGroup().addTo(this.map);

    for (let item of coordinates) {
      leaflet.marker(item, {icon: this.icon})
        .addTo(this.group);
    }
    leaflet.marker(selectedOffer, {icon: this.activeIcon}).addTo(this.group);
  }

  _initMap() {
    const city = CITIES.get(this.props.currentCity);

    this.map = leaflet.map(MapConfig.ID, {
      center: city,
      zoom: MapConfig.ZOOM,
      zoomControl: false,
      // marker: true,
      scrollWheelZoom: false,
    });

    this.map.setView(city, MapConfig.ZOOM);

    leaflet
      .tileLayer(MapConfig.TITLE_LAYER)
      .addTo(this.map);
    this._addMarkers();
    this.map.on(`click`, () => {
      if (this.map.scrollWheelZoom.enabled()) {
        this.map.scrollWheelZoom.disable();
      } else {
        this.map.scrollWheelZoom.enable();
      }
    });
  }

  componentDidMount() {
    this._initMap();
  }

  componentDidUpdate() {
    const {currentCity, coordinates, selectedOffer} = this.props;

    this.group.clearLayers();
    this.map.setView(CITIES.get(currentCity), MapConfig.ZOOM);

    for (let item of coordinates) {
      leaflet.marker(item, {icon: this.icon})
        .addTo(this.group);
    }

    leaflet.marker(selectedOffer, {icon: this.activeIcon})
      .addTo(this.group);
  }


  render() {
    return (
      <div className="cities__map" id="map"/>
    );
  }
}


export default Map;
