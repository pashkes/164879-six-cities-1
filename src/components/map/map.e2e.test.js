import React from 'react';
import {Map} from './map.jsx';
import renderer from 'react-test-renderer';

const coordinates = [[52.3809553943508, 4.939309666406198]];

it(`map render is correctly`, () => {
  const map = renderer
    .create(<Map currentCity={`Amsterdam`} coordinates={coordinates}/>)
    .toJSON();
  expect(map).toMatchSnapshot();
});
