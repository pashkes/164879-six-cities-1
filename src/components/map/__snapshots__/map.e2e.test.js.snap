// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map render is correctly 1`] = `
<div
  className="cities__map"
  id="map"
/>
`;
